MdePkg/Library/UefiUsbLib/UsbDxeLib.c

```diff
@@ -4,12 +4,16 @@
   in Usb specification 9.4 section.
 
   Copyright (c) 2004 - 2018, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) 2024, American Megatrends International LLC. All rights reserved.<BR>
   SPDX-License-Identifier: BSD-2-Clause-Patent
 
 **/
 
 #include "UefiUsbLibInternal.h"
 
+STATIC UINT8                      *mConfigData;
+STATIC EFI_USB_DEVICE_DESCRIPTOR  mDeviceDescriptor;

```

Comment:
Use of global variables that hold state info can cause problems if APIs are called with different USB I/O Instances.

Reply:
Global variables contain the state info of the last call. They are validated on every function entry, and updated if USB I/O instance is different. This minimizes HW access to the device.

Reply:
What happens if 2 USB drivers with timer events use these services.  Can they get confused?  How does the caller know it needs to start over?

Reply:
Every caller passes its UsbIo, this instance is validated and used within a function. InitUsbConfigDescriptorData checks if the current data belong to the given UsbIo instance. If not, the given UsbIo instance is used to retrieve the new data.
Not sure I got the failed scenario.

Reply:
Ok.  As long as the globals are only used as a cache for performance it is fairly safe,  There are still cases where this can fail.  If 2 USB drivers use timers at different TPLs, then one timer event can call an API and update the cache.  In the middle of that API call, a timer event at a higher TPL for a different USB driver can call an API and update the cache for the 2nd USB driver and will complete without issue.  Then the first timer event will resume with the cached content from the 2nd USB driver without knowing the cache was replaced with a different device.

Reply:
I agree. Do you think raising TPL to TPL_NOTIFY during the calls would resolve this?

Reply:
Yes.  I think raising to TPL_NOTIFY in all APIs that get/set the global variables will resolve this issue.

Reply:
done